File: game/lvl.lua
local stream = require("stream")
local console = require("console")
local config = require("config")

lvl = {}

function lvl.load(address)
    lvl.offset = address
	local stream = stream.open(address)

    console.log("pink", "Reading LVL @ %X", lvl.offset)

    stream:advance(4 * 4)
    lvl.timestamp = stream:readstring(0x18)
    stream:advance(0xf0)
    lvl.numTextures = stream:read32()

    if config.PRINT_INFO then
        console.log("grey", "Number of textures: %d", lvl.numTextures)
    end

    local totalTextureCount = lvl.numTextures - fix.numTextures
    stream:advance(totalTextureCount * 4) -- skip texture pointers
    stream:advance(totalTextureCount * 4) -- skip texture pointers

    -- Most of these pointers seem to be invalid
    lvl.actualWorld = stream:readpointer()
    lvl.dynamicWorld = stream:readpointer()
    lvl.inactiveDynamicWorld = stream:readpointer()
    lvl.fatherSector = stream:readpointer()
    lvl.submapPos = stream:readpointer()
    lvl.numAlways = stream:read32()
    stream:advance(4 * 3) -- TODO: read spawnable persos
    stream:advance(4 * 3) -- skip for now

    if config.PRINT_INFO then
        console.log("green", "Reading lvl object names @ %X", stream.position)
    end

    lvl.objectTypes = {}
    for i = 0, 2 do
        local nameFirst = stream:readpointer()
        local nameLast = stream:readpointer()
        local numNames = stream:read32()

        --print("Num names: " .. numNames)

        if numNames > 10000 then
            console.log("red", "Unreasonable number of names - invalid LVL!")
            return nil
        end

        lvl.objectTypes[i] = {}
        local current = nameFirst

        if config.PRINT_INFO then
            local group = {"Family", "Model", "Object"}
            console.log("grey", "\t[%s]", group[i+1])
        end

        local s = stream.open(current)
        for n = 0, numNames-1 do

            lvl.objectTypes[i][n] = {}
            lvl.objectTypes[i][n].namesNextOffset = s:readpointer()
            lvl.objectTypes[i][n].namesPrevOffset = s:readpointer()
            lvl.objectTypes[i][n].headerOffset = s:readpointer()
            lvl.objectTypes[i][n].nameOffset = s:readpointer()

            lvl.objectTypes[i][n].name = memory.readstring(lvl.objectTypes[i][n].nameOffset)
            --print("name: " .. lvl.objectTypes[i][n].name)

            if config.PRINT_INFO then
                local a = console.fmt("pink", "\t[lvl @ %X]: ", s.position)
        		local b = console.fmt("blue", "%s", lvl.objectTypes[i][n].name)
        		print(a .. b)
            end

            if lvl.objectTypes[i][n].namesNextOffset ~= 0x00 then
                s:jump(lvl.objectTypes[i][n].namesNextOffset)
            end
        end

        --print ""
    end

    console.log("yellow", "Finished reading LVL\n")
end

File: game/global/lvl.c
//
//  lvl.c
//  r3lib
//
//  Created by Jba03 on 2022-11-21.
//

#include "lvl.h"
#include "fix.h"
#include "global.h"

#include <stdio.h>
#include <stdlib.h>

LVL_STRUCTURE struct LVL* lvl_read(const address addr)
{
    struct LVL* lvl = malloc(sizeof *lvl);
    lvl->offset = addr;
    
    struct Stream *stream = stream_open(addr);
    advance(4 * 4);
    strcpy(lvl->timestamp, readstring(0x18));
    advance(0xF0);
    lvl->texture_count = read32();
    
    uint32_t total_texture_count = lvl->texture_count - fix->texture_count;
    advance(total_texture_count * 4); /* Skip texture pointers */
    advance(total_texture_count * 4); /* Skip texture pointers */
    
    lvl->actual_world = readpointer();
    lvl->dynamic_world = readpointer();
    lvl->inactive_dynamic_world = readpointer();
    lvl->father_sector = readpointer();
    lvl->submap_position = readpointer();
    lvl->num_always = read32();
    
    advance(4 * 3); /* Spawnable persos */
    advance(4 * 3);
    
    for (int i = 0; i < 3; i++)
    {
        uint32_t checkpoint = stream->position;
        pointer name_first = readpointer();
        pointer name_last = readpointer();
        uint32_t n_names = read32();
        
        info(COLOR_GREEN "\t[%s names]\n", (const char* []){"Family", "Model", "Instance"}[i]);
        
        stream_seek(stream, name_first);
        for (int n = 0; n < n_names; n++)
        {
            pointer name_next = readpointer();
            pointer name_prev = readpointer();
            pointer header_offset = read32();
            pointer name_offset = readpointer();
            
            const char* cur;
            stream_seek(stream, name_offset);
            switch (i)
            {
                case 0: cur = lvl->family_names[n] = readstringN(); family_color_table[n] = (rand() % 0xFFFFFF) | 0xFF000000; break;
                case 1: cur = lvl->model_names[n] = readstringN(); break;
                case 2: cur = lvl->instance_names[n] = readstringN(); break;
                default: cur =  lvl->instance_names[n] = "Invalid name";
            }
            
            info(COLOR_PINK "\t%s\n", cur);
            
            if (name_next == 0x00) break;
            stream_seek(stream, name_next);
        }
        stream_seek(stream, checkpoint + 3 * 4);
    }
    
    stream_close(stream);
    
    return lvl;
}
